### DISCUSSION TITLE
How to turn off sqlite json query with JSON_QUOTE? There are different between SQLAlchemy's sqlite and mysql json query.

### DISCUSSION THREAD

**Discussion by vba34520** [sqlalchemy/sqlalchemy#9480] [Usage Questions] [answered]
**MySQL**

JSON query statement is clear and beautiful like this.

```python
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import create_engine, Column, Integer, JSON

Base = declarative_base()


class Person(Base):
    __tablename__ = 'person'

    id = Column(Integer(), primary_key=True)
    message = Column(JSON)


engine = create_engine('mysql+pymysql://root:123456@localhost:3306/test', pool_recycle=3600)  # MySQL
Base.metadata.create_all(engine)
Session = sessionmaker(bind=engine)
session = Session()

person = Person(message={'name': 'Bob'})  # Insert a record
session.add(person)
session.commit()

query = session.query(Person.id).filter(Person.message['name'] == 'Bob')  # MySQL will work, clear and beautiful
print(query[0].id)  # 1
print(query)
# SELECT person.id AS person_id 
# FROM person 
# WHERE JSON_EXTRACT(person.message, %(message_1)s) = %(param_1)s
```

**SQLite**

It uses needless `json_quote()` and the result doesn't clear. I want to turn off sqlite json query with JSON_QUOTE.

How can I solve this?

```python
from sqlalchemy import func
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import create_engine, Column, Integer, JSON

Base = declarative_base()


class Person(Base):
    __tablename__ = 'person'

    id = Column(Integer(), primary_key=True)
    message = Column(JSON)


engine = create_engine('sqlite:///:memory:')  # SQLite
Base.metadata.create_all(engine)
Session = sessionmaker(bind=engine)
session = Session()

person = Person(message={'name': 'Bob'})  # Insert a record
session.add(person)
session.commit()

query = session.query(Person.id).filter(Person.message['name'] == func.json_quote('Bob'))  # SQLite will work, but need to use json_quote
print(query[0].id)  # 1
print(query)
# SELECT person.id AS person_id 
# FROM person 
# WHERE JSON_QUOTE(JSON_EXTRACT(person.message, ?)) = json_quote(?)
```

**Comment by zzzeek** [comment, marked as answer]
hi - 

take a look at "Casting JSON Elements to other types" at https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.JSON .   basically `struct["key"]` can be any JSON, and in the case of MySQL, it's working by accident.  the correct, platform agnostic way to compare an element in a json struct as a string to a string is to use as_string(); otherwise SQLAlchemy doesn't know what kind of comparison you want to do:


    query = session.query(Person.id).filter(
        Person.message["name"].as_string() == "Bob"
    )

**Reply by vba34520** [reply]
Is it possible to turn off sqlite json query with JSON_QUOTE?

Because I'm writing tests code in sqlite and there are a lot of business code using mysql.

**Reply by zzzeek** [reply]
there's not really a way to change this operator just for SQLite and not all other dialects, this would involve creating new JSON comparison objects and new compilation rules, it would be very complicated.

What you should do is actually fix your program; run your test suite against SQLite, and all those areas that fail, simply add the required as_string() call.  MySQL will be unaffected.
